src/finch/symbolic/term.py

```diff
@@ -1,56 +1,65 @@
-from typing import Any, Iterator
-from abc import ABC, abstractmethod
-
 """
-This module contains definitions for common functions that are useful for symbolic expression manipulation.
-Its purpose is to provide a shared interface between various symbolic programming in Finch.
+This module contains definitions for common functions that are useful for symbolic
+expression manipulation. Its purpose is to provide a shared interface between various
+symbolic programming in Finch.
 
 Classes:
-    Term (ABC): An abstract base class representing a symbolic term. It provides methods to access the head
-    of the term, its children, and to construct a new term with a similar structure.
+    Term (ABC): An abstract base class representing a symbolic term. It provides methods
+    to access the head of the term, its children, and to construct a new term with a
+    similar structure.
 """
 
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from collections.abc import Iterator
+from typing import TYPE_CHECKING, Self
+
+__all__ = ["Term", "PreOrderDFS", "PostOrderDFS"]
+
+if TYPE_CHECKING:
+    from ..finch_logic import LogicNode
+
 
 class Term(ABC):
     def __init__(self):
         self._hashcache = None  # Private field to cache the hash value
 
     @abstractmethod
-    def head(self) -> Any:
-        """Return the head type of the S-expression."""
-        pass
-
-    def children(self) -> list["Term"]:
+    def children(self) -> list[LogicNode]:
         """Return the children (AKA tail) of the S-expression."""
-        pass
 
     @abstractmethod
     def is_expr(self) -> bool:
-        """Return True if the term is an expression tree, False otherwise. Must implement children() if True."""
-        pass
+        """
+        Return True if the term is an expression tree, False otherwise. Must implement
+        `children()` if `True`."""
 
     @abstractmethod
-    def make_term(self, head: Any, children: list["Term"]) -> "Term":
+    def make_term(self, *children: Term) -> Self:
         """
-        Construct a new term in the same family of terms with the given head type and children.
-        This function should satisfy `x == x.make_term(x.head(), *x.children())`
+        Construct a new term in the same family of terms with the given
+        children. This function should satisfy
+        `x == x.make_term(*x.children())`
         """
-        pass
 
     def __hash__(self) -> int:
         """Return the hash value of the term."""
         if self._hashcache is None:
             self._hashcache = hash(
-                (0x1CA5C2ADCA744860, self.head(), tuple(self.children()))
+                (0x1CA5C2ADCA744860, type(self), tuple(self.children()))
```

Comment:
Let's keep head.

Reply:
Done in https://github.com/finch-tensor/finch-tensor-lite/pull/17/commits/56349b0d36760ac0f6f916d9b851a2d6744caddd.